rbs/TypeTranslator.cc

```diff
@@ -0,0 +1,424 @@
+                keysStore.emplace_back(ast::MK::Symbol(loc, keyName));
+                break;
+            }
+            case RBS_AST_STRING: {
+                rbs_ast_string_t *keyNode = (rbs_ast_string_t *)hash_node->key;
+                keyStr = std::string(keyNode->string.start);
+                auto keyName = ctx.state.enterNameUTF8(keyStr);
+                keysStore.emplace_back(ast::MK::String(loc, keyName));
+                break;
+            }
+            default: {
+                if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSError)) {
+                    e.setHeader("Unexpected node type `{}` in record key type, expected `{}`",
+                                rbs_node_type_name(hash_node->key), "Symbol");
+                }
+                continue;
+            }
+        }
+
+        if (hash_node->value->type != RBS_TYPES_RECORD_FIELDTYPE) {
+            if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSError)) {
+                e.setHeader("Unexpected node type `{}` in record value type, expected `{}`",
+                            rbs_node_type_name(hash_node->value), "RecordFieldtype");
+            }
+
+            continue;
+        }
+
+        rbs_types_record_fieldtype_t *valueNode = (rbs_types_record_fieldtype_t *)hash_node->value;
+        auto innerType = TypeTranslator::toRBI(ctx, typeParams, valueNode->type, loc);
+        valuesStore.emplace_back(std::move(innerType));
+    }
+
+    return ast::MK::Hash(loc, std::move(keysStore), std::move(valuesStore));
+}
+
+sorbet::ast::ExpressionPtr variableType(core::MutableContext ctx, rbs_types_variable_t *node, core::LocOffsets loc) {
+    rbs_ast_symbol_t *symbol = (rbs_ast_symbol_t *)node->name;
+    rbs_constant_t *constant = rbs_constant_pool_id_to_constant(fake_constant_pool, symbol->constant_id);
+    std::string string(constant->start, constant->length);
+    auto name = ctx.state.enterNameUTF8(string);
+    return ast::MK::Send1(loc, ast::MK::T(loc), core::Names::typeParameter(), loc, ast::MK::Symbol(loc, name));
+}
+
+} // namespace
+
+core::LocOffsets TypeTranslator::nodeLoc(core::LocOffsets offset, rbs_node_t *node) {
+    auto loc = node->location;
+    return core::LocOffsets{
+        offset.beginPos() + loc->rg.start.char_pos + 2,
+        offset.beginPos() + loc->rg.end.char_pos + 2,
+    };
+}
+
+sorbet::ast::ExpressionPtr TypeTranslator::toRBI(core::MutableContext ctx,
+                                                 std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                                                 rbs_node_t *node, core::LocOffsets docLoc) {
+    switch (node->type) {
+        case RBS_TYPES_ALIAS: {
+            auto loc = TypeTranslator::nodeLoc(docLoc, node);
+            if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSError)) {
+                e.setHeader("RBS aliases are not supported yet");
+            }
```

Comment:
I don't know what the plan is for supporting the unsupported bits here, but perhaps the unsupported cases should have a separate error code that we can eventually retire?

Reply:
I am inclined to say that `RBSInternalError` should be `ENFORCE`s instead.  We do have `sorbet::core::errors::Internal::InternalError`, which I guess we could use for this?  But scanning through uses of `InternalError`, we are typically throwing an exception along with reporting an error, and I think we'd want to do the same if we used `InternalError` here.

Reply:
I removed the `RBSInternalError` and replaced them by `ENFORCE` (and `InternalError` in a few places) 👍